### PR TITLE
Start remediate action for hosts via bulk action

### LINKED ISSUE #73: start remediate actions via API call
It should be possible to start a remediate action for a host and status via API.

### LINKED ISSUE #74: start remediate action for hosts via bulk action
It should be possible to bulk select hosts and schedule a remediate action for them.

### PR DESCRIPTION
**Pull request by kamils-iRonin**
closes #74

With "Remediate all" you can start remediate action for all non-ok statuses.
image

You can also select the statuses for which you want to start remediate action.
image

### CODE REVIEW

File: app/controllers/foreman_wreckingball/hosts_controller.rb [hunk 1]
```diff
@@ -112,6 +134,10 @@ def find_status
       @status = HostStatus::Status.find_by!(:id => params[:status_id], :host_id => @host.id)
     end
 
+    def find_statuses
+      @statuses = HostStatus::Status.includes(:host).find(params[:status_ids])
```
Comment (timogoebel) [inline app/controllers/foreman_wreckingball/hosts_controller.rb]: Does this properly check permissions?

Comment (kamils-iRonin) [reply app/controllers/foreman_wreckingball/hosts_controller.rb]: checking permissions is not implemented yet, I'll add it with proper test

**Comment by kamils-iRonin** [comment]
@timogoebel Can you please take a look? I've moved triggering remediate actions to the seperate [job](https://github.com/dm-drogeriemarkt/foreman_wreckingball/pull/78/files#diff-fc57e19e41fbdc8e54c54e9c27313d3aR6). I did some tests and it seems to work fine. WDYT about it?
Tests are missing

File: app/views/foreman_wreckingball/hosts/_status_row_actions.html.erb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<% actions = [] %>
+<% actions << display_link_if_authorized(_('Refresh'), hash_for_refresh_status_dashboard_hosts_path, title: _('Refresh data'), method: :put) %>
```
Comment (timogoebel) [inline app/views/foreman_wreckingball/hosts/_status_row_actions.html.erb]: It looks like this should be moved to a helper (method).

File: app/controllers/foreman_wreckingball/hosts_controller.rb [hunk 3]
```diff
@@ -108,8 +108,27 @@ def triggering_params
...
 
-    def find_status
-      @status = HostStatus::Status.find_by!(:id => params[:status_id], :host_id => @host.id)
+    def statuses_params
+      @statuses_params ||= params.permit(:host_association, :owned_only, status_ids: [])
+    end
+
+    def find_statuses
+      @statuses = begin
+        host_association = statuses_params[:host_association]&.to_sym
```
Comment (timogoebel) [inline app/controllers/foreman_wreckingball/hosts_controller.rb]: "Avoid usage of to_sym on user-supplied input."
https://www.sitepoint.com/rails-security-pitfalls/

Does this apply here?

File: app/views/foreman_wreckingball/hosts/schedule_remediate.html.erb [hunk 4]
```diff
@@ -1,35 +1,52 @@
...
 
+<% if @statuses.any? %>
+  <% if @statuses.group_by(&:class).map { |i| i.first}.select { |i| i.try(:dangerous_remediate?) }.any? %>
+    <%= alert(:text => _('This will cause a service interruption.'), :class => 'alert-warning', :close => false) %>
+  <% end %>
 
-<% if @status.class.respond_to?(:dangerous_remediate?) && @status.class.dangerous_remediate? -%>
-<%= alert(:text => _('This will cause a service interruption.'), :class => 'alert-warning', :close => false) %>
-<% end -%>
+  <%= _('%s hosts selected for remedate' % @statuses.count) %>
```
Comment (timogoebel) [inline app/views/foreman_wreckingball/hosts/schedule_remediate.html.erb]: Typo: `remediate`, but i think it should be `for remediation.`

@leewaa: Can you help?

Comment (timogoebel) [reply app/views/foreman_wreckingball/hosts/schedule_remediate.html.erb]: Also: Please use `n_("One", "Two", number)` so plural forms are matched correctly.

File: app/views/foreman_wreckingball/hosts/_status_row_actions.html.erb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+<% actions << display_link_if_authorized(_('Remediate all'), hash_for_schedule_remediate_hosts_path, 'data-host-association': host_association, 'data-owned-only': owned_only, onClick: "show_modal(this); return false;") if supports_remediate %>
```
Comment (timogoebel) [inline app/views/foreman_wreckingball/hosts/_status_row_actions.html.erb]: Nit: I would capitalize this. `Remediate All`

**Comment by leewaa** [comment]
Looks good, only one small thing.

When navigating `Back` to the status page after having checked all vms in a list, then the checkbox is still checked with nothing selected:

image

**Comment by kamils-iRonin** [comment]
> Looks good, only one small thing.
> 
> When navigating `Back` to the status page after having checked all vms in a list, then the checkbox is still checked with nothing selected

@leewaa Thanks for testing. Fixed.

File: app/models/concerns/foreman_wreckingball/host_status_extensions.rb [hunk 6]
```diff
@@ -7,7 +7,7 @@ def wreckingball_statuses
     end
 
     def find_wreckingball_status_by_host_association(host_association)
-      wreckingball_statuses.find { |s| s.host_association == host_association }
+      wreckingball_statuses.find { |s| s.host_association.to_s == host_association.to_s }
```
Comment (timogoebel) [inline app/models/concerns/foreman_wreckingball/host_status_extensions.rb]: Can we do this via active record, e.g. via a custom scope?

File: app/controllers/foreman_wreckingball/hosts_controller.rb [hunk 7]
```diff
@@ -108,8 +108,27 @@ def triggering_params
...
+        host_association = statuses_params[:host_association]
+        status_class = HostStatus.find_wreckingball_status_by_host_association(host_association)
+        if status_class
+          Host.authorized(:remediate_vmware_status_hosts, Host)
+              .joins(status_class.host_association)
+              .includes(status_class.host_association)
+              .try { |query| statuses_params[:owned_only] ? query.owned_by_current_user_or_group_with_current_user : query }
+              .where.not('host_status.status': status_class.global_ok_list)
+              .map { |host| host.send(status_class.host_association) }
+        else
```
Comment (timogoebel) [inline app/controllers/foreman_wreckingball/hosts_controller.rb]: Why do we need the else branch here?

Comment (kamils-iRonin) [reply app/controllers/foreman_wreckingball/hosts_controller.rb]: There are two ways to start remediate action: you can select hosts for remediation and then `status_ids` is passed in the params or you can remediate all hosts for status and then `host_association` and optionally `owner_only` are passed in the params.

File: app/models/concerns/foreman_wreckingball/host_status_extensions.rb (hunk 6 above)
Comment (kamils-iRonin) [reply app/models/concerns/foreman_wreckingball/host_status_extensions.rb]: I don't think we can use custom scope here. We have to go through statuses registry and find wreckingball status that match `host_association`.

**Comment by kamils-iRonin** [comment]
> Also this causes lock conflicts

@timogoebel I spent a lot of time trying to reproduce it but without success. Can you please test it again and provide more detailed description?

> The title of the page shows errors:

The page title also looks ok to me

**Comment by timogoebel** [comment]
Thanks, @kamils-iRonin.

**Pull request merged**